packages/objectloader2/src/operations/database.ts

```diff
@@ -0,0 +1,198 @@
+        }
+      }
+
+      request.onsuccess = () => resolve(request.result)
+      request.onerror = (e) =>
+        reject(ensureError(e, 'Failed to open IndexedDB database'))
+    })
+  }
+
+  private supportsCache(): boolean {
+    return !!(this._options.enableCaching && this._options.indexedDB)
+  }
+
+  private async setupCacheDb(): Promise<boolean> {
+    if (this._cacheDB !== undefined && !this.supportsCache()) return false
+
+    // Initialize
+    await this.safariFix()
+    this._cacheDB = await this.openDatabase(
+      CacheDatabase._databaseName,
+      CacheDatabase._storeName
+    )
+    return true
+  }
+  async getItems(
+    baseIds: string[],
+    found: Queue<Item>,
+    notFound: Queue<string>
+  ): Promise<void> {
+    if (!(await this.setupCacheDb())) {
+      return
+    }
+
+    let count = 0
+    for (let i = 0; i < baseIds.length; i += 500) {
+      const baseIdsChunk = baseIds.slice(i, i + 500)
+
+      const store = this._cacheDB!.transaction(
+        CacheDatabase._storeName,
+        'readonly'
+      ).objectStore(CacheDatabase._storeName)
+      const idbChildrenPromises = baseIdsChunk.map<Promise<void>>(async (baseId) => {
+        const getBase = new Promise((resolve, reject) => {
+          const request = store.get(baseId)
+
+          request.onsuccess = () => resolve(request.result)
+          request.onerror = () =>
+            reject(ensureError(request.error, 'Error trying to get a batch'))
+        })
+        const base = await getBase
+        count++
+        if (base === undefined) {
+          notFound.add(baseId)
+          if (count % 1000 === 0) {
+            this._logger(`Object ${count} not found in cache`)
+          }
+        } else {
+          if (count % 1000 === 0) {
+            this._logger(`Object ${count} found in cache`)
+          }
+          if (isBase(base)) {
+            found.add({ baseId, base })
+          } else {
+            throw new ObjectLoaderRuntimeError(`${baseId} is not a base`)
+          }
+        }
+      })
+      await Promise.all(idbChildrenPromises)
+    }
+  }
+
+  async getItem(baseId: string): Promise<Item | undefined> {
+    if (!(await this.setupCacheDb())) {
+      return undefined
+    }
+
+    const store = this._cacheDB!.transaction(
+      CacheDatabase._storeName,
+      'readonly'
+    ).objectStore(CacheDatabase._storeName)
+    const getBase = new Promise<unknown>((resolve, reject) => {
+      const request = store.get(baseId)
+
+      request.onsuccess = () => resolve(request.result)
+      request.onerror = () =>
+        reject(ensureError(request.error, 'Error trying to get an item'))
+    })
+    const base = await getBase
+    if (base === undefined) return undefined
+    if (isBase(base)) {
+      return { baseId, base }
+    } else {
+      throw new ObjectLoaderRuntimeError(`${baseId} is not a base`)
+    }
+  }
+
+  static async cacheSaveBatch(batch: Item[], cacheDB: IDBDatabase): Promise<void> {
```

Comment:
why are these 2 functions static while everything else isn't?

Reply:
usually I get suggestions to have things as static when they are but since you have to address static members directly then it's uglier